src/client-side-metrics/operation-metrics-collector.ts

```diff
@@ -0,0 +1,356 @@
+   */
+  onResponse(projectId: string) {
+    if (!this.receivedFirstResponse) {
+      this.receivedFirstResponse = true;
+      const endTime = this.dateProvider.getDate();
+      if (projectId && this.operationStartTime) {
+        this.firstResponseLatency =
+          endTime.getTime() - this.operationStartTime.getTime();
+      }
+    }
+  }
+
+  /**
+   * Called when an operation completes (successfully or unsuccessfully).
+   * Records operation latencies, retry counts, and connectivity error counts.
+   * @param {string} projectId The id of the project.
+   * @param {OperationOnlyAttributes} info Information about the completed operation.
+   */
+  onOperationComplete(projectId: string, info: OperationOnlyAttributes) {
+    if (
+      this.state ===
+      MetricsCollectorState.OPERATION_STARTED_ATTEMPT_NOT_IN_PROGRESS
+    ) {
+      this.state = MetricsCollectorState.OPERATION_COMPLETE;
+      const endTime = this.dateProvider.getDate();
+      if (projectId && this.operationStartTime) {
+        const totalTime = endTime.getTime() - this.operationStartTime.getTime();
+        {
+          const operationAttributes = {
+            finalOperationStatus: info.finalOperationStatus,
+            streamingOperation: info.streamingOperation,
+            projectId,
+            instanceId: this.tabularApiSurface.instance.id,
+            table: this.tabularApiSurface.id,
+            cluster: this.cluster,
+            zone: this.zone,
+            appProfileId: this.tabularApiSurface.bigtable.appProfileId,
+            methodName: this.methodName,
+            clientName: `nodejs-bigtable/${version}`,
+          };
+          const metrics = {
+            operationLatency: totalTime,
+            retryCount: this.attemptCount - 1,
+            firstResponseLatency: this.firstResponseLatency ?? undefined,
+          };
+          this.metricsHandlers.forEach(metricsHandler => {
+            if (metricsHandler.onOperationComplete) {
+              metricsHandler.onOperationComplete(metrics, operationAttributes);
+            }
+          });
+        }
+      }
+    } else {
+      console.warn('Invalid state transition attempted');
+    }
+  }
+
+  /**
+   * Called when metadata is received. Extracts server timing information if available.
+   * @param {string} projectId The id of the project.
+   * @param {object} metadata The received metadata.
+   */
+  onMetadataReceived(
+    projectId: string,
+    metadata: {
+      internalRepr: Map<string, Buffer>;
+      options: {};
+    }
+  ) {
+    const mappedEntries = new Map(
+      Array.from(metadata.internalRepr.entries(), ([key, value]) => [
+        key,
+        value.toString(),
+      ])
+    );
+    const durationValues = mappedEntries.get('server-timing')?.split('dur=');
+    if (durationValues && durationValues[1]) {
+      if (!this.serverTimeRead) {
+        this.serverTimeRead = true;
+        const serverTime = parseInt(durationValues[1]);
+        if (projectId) {
```

Comment:
why check the projectID way down here?

Reply:
I removed projectId from here. This must be a leftover artifact.